### PR TITLE
feat: rate limit password reset attempts

### PR DESCRIPTION
**Pull request by kwasniew**
## About the changes

* added server side rate limiter (3 resets per minute)
* improved UI to handle 429 code
* refactored UI to handle exclusive states of the UI and to make impossible states impossible

Screenshot 2024-02-16 at 13 12 19

### CODE REVIEW

**Review by ivarconr — APPROVED** [review]

File: src/lib/features/metrics/client-metrics/client-metrics.e2e.test.ts [hunk 1]
```diff
@@ -162,6 +162,7 @@ test('should support the hoursBack query param for raw metrics', async () => {
     expect(hoursTooFew.data).toHaveLength(24);
     expect(hoursTooMany.data).toHaveLength(24);
     expect(days.data).toHaveLength(2); // two days of data
+    console.log(days);
```
Comment (thomasheartman) [inline src/lib/features/metrics/client-metrics/client-metrics.e2e.test.ts:165]: Is this left in on purpose?

File: frontend/src/component/user/ForgottenPassword/ForgottenPassword.tsx [hunk 2]
```diff
@@ -48,37 +48,40 @@ const StyledTypography = styled(Typography)(({ theme }) => ({
     ...textCenter,
 }));
 
+type State = 'initial' | 'loading' | 'attempted' | 'too_many_attempts';
```
Comment (thomasheartman) [inline frontend/src/component/user/ForgottenPassword/ForgottenPassword.tsx:51]: I like this a lot 😄

**Review by thomasheartman — APPROVED** [review]
Nice work 😄

File: src/lib/features/metrics/client-metrics/client-metrics.e2e.test.ts (hunk 1 above)
Comment (kwasniew) [reply src/lib/features/metrics/client-metrics/client-metrics.e2e.test.ts]: thanks for catching this. my mistake

**Pull request merged**